recipes/CommonVoice/self-supervised-learning/wav2vec2/train.py

```diff
@@ -0,0 +1,294 @@
+
+
+# Define training procedure
+class W2VBrain(sb.core.Brain):
+    def compute_forward(self, batch, stage):
+        """Forward computations from the waveform batches to the w2v2 loss."""
+
+        batch = batch.to(self.device)
+        wavs, wav_lens = batch.sig
+        wavs, wav_lens = wavs.to(self.device), wav_lens.to(self.device)
+
+        # Normalize wav signal
+        wavs = self.modules.normalize(wavs, wav_lens)
+
+        # Forward on w2v2 and take the loss.
+        out, mask = self.modules.wav2vec2(wavs, mask_device=self.device)
+        loss = out.loss
+
+        if stage != sb.Stage.TRAIN:
+            with torch.no_grad():
+                out, mask = self.modules.wav2vec2(wavs, mask_device=self.device)
+
+            return out.loss, out, mask
+
+        return loss
+
+    def compute_objectives(self, predictions, batch, stage):
+        """Computes the loss (CTC+NLL) given predictions and targets."""
+
+        if stage == sb.Stage.TRAIN:
+            # We don't have to compute anything as the HF model directly returns
+            # the constrative loss.
+            loss = predictions
+        else:
+            # We compute the accuracy between embeddings with cosing sim.
+            loss, out, mask_time_indices = predictions
+            cosine_sim = torch.cosine_similarity(
+                out.projected_states, out.projected_quantized_states, dim=-1
+            )
+            acc = cosine_sim[mask_time_indices].mean()
+            self.acc_metric.append(acc)
+
+        return loss
+
+    def fit_batch(self, batch):
+        """Train the parameters given a single batch in input"""
+        predictions = self.compute_forward(batch, sb.Stage.TRAIN)
+        loss = self.compute_objectives(predictions, batch, sb.Stage.TRAIN)
+
+        # normalize the loss by gradient_accumulation step
+        (loss / self.hparams.gradient_accumulation).backward()
+
+        if self.step % self.hparams.gradient_accumulation == 0:
+            # gradient clipping & early stop if loss is not fini
+            self.check_gradients(loss)
+
+            self.optimizer.step()
+            self.optimizer.zero_grad()
+
+            # anneal lr every update
+            self.hparams.noam_annealing(self.optimizer)
+
+        return loss.detach()
+
+    def on_stage_start(self, stage, epoch):
+        """Gets called at the beginning of each epoch"""
+        if stage != sb.Stage.TRAIN:
+            self.acc_metric = []
+
+    def on_stage_end(self, stage, stage_loss, epoch):
+        """Gets called at the end of an epoch."""
+        # Compute/store important stats
+        stage_stats = {"loss": stage_loss}
+        if stage == sb.Stage.TRAIN:
+            self.train_stats = stage_stats
+        else:
+            stage_stats["acc"] = sum(self.acc_metric) / len(self.acc_metric)
+
+        # Perform end-of-iteration things, like annealing, logging, etc.
+        if stage == sb.Stage.VALID and sb.utils.distributed.if_main_process():
```

Comment:
@TParcollet , do you have some bugs here ?
in our `core.py` we have already have the call of `run_on_main`:
https://github.com/speechbrain/speechbrain/blob/68a8b5979dc5d6a885489d9a0428682d5a2ac012/speechbrain/core.py#L1078
let me do a fix of the use of `local_rank` and `rank` in our distributed solution

Reply:
I guess we can remove it, dunno why this is added here.